torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -150,26 +160,38 @@ class NCCLComm {
   }
 
   ncclComm_t getNcclComm() {
-    std::unique_lock<std::mutex> lock(mutex_);
     if (aborted_) {
-      TORCH_CHECK(false,
-          "NCCL communicator was aborted on rank " + std::to_string(rank_) +
-          ".");
+      std::string abortMsg = "NCCL communicator was aborted on rank " + std::to_string(rank_) + ".";
+      if (commFailureReason_ != c10::nullopt) {
+        abortMsg += " . Original reason for failure was: " + *commFailureReason_;
+      }
+      TORCH_CHECK(false, abortMsg);
     }
     return ncclComm_;
   }
 
-  void ncclCommAbort() {
+  c10::optional<std::string> getNcclCommFailureReason() const {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return commFailureReason_;
+  }
+
+  void ncclCommAbort(
+      c10::optional<std::string> commFailureReason = c10::nullopt) {
     std::unique_lock<std::mutex> lock(mutex_);
 #ifdef ENABLE_NCCL_ERROR_CHECKING
     if (aborted_) {
       // Should not abort twice.
       return;
     }
 
-    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_));
+    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_), commFailureReason_);
     aborted_ = true;
     ncclComm_ = nullptr;
+    // Set true failure reason if provided by ProcessGroupNCCL (e.g. work
+    // timeout)
+    if (commFailureReason != c10::nullopt) {
+      commFailureReason_ = commFailureReason;
```

Comment:
nit: We can just do `commFailureReason_ = commFailureReason` without the conditional?